File: haskell/interpreter/src/Token.hs
module Token (TokenType (..)) where

import Data.Int (Int64)

data TokenType
    = Illegal String
    | Eof
    | -- Idenfitieres
      Ident String
    | Int Int64
    | Str String
    | -- Operators
      Assign
    | Plus
    | Minus
    | Bang
    | Asterisk
    | Slash
    | Lt
    | Gt
    | Eq
    | NotEq
    | -- Delimeters
      Comma
    | Colon
    | Semicolon
    | LParen
    | RParen
    | LBrace
    | RBrace
    | LBracket
    | RBracket
    | -- Keywords
      Function
    | Let
    | KTrue
    | KFalse
    | If
    | Else
    | Return
    deriving (Show, Eq)

File: rust/interpreter/src/token.rs
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Token<'input> {
    Illegal(u8),
    Eof,

    // Identifierts + literals
    Ident(&'input str),
    Int(i64),

    // Operators
    Assign,
    Plus,
    Minus,
    Bang,
    Asterisk,
    Slash,

    Lt,
    Gt,

    Eq,
    Neq,

    // Delimiters
    Comma,
    Semicolon,

    LBrace,
    RBrace,
    LParen,
    RParen,
    
    // keywords
    Fun,
    Let,
    False,
    True,
    Return,
    If,
    Else,
}
